### ISSUE TITLE
No video until power cycle of game [linux - nVidia-specific?]

### ISSUE THREAD

**Issue by McKayJT** [SourMesen/Mesen#319]
After a new rom is loaded, there is no video of the game (audio works) until the game is power cycled. Reset doesn't work to fix this; it must be a hard power cycle. After you do this, it displays normally until a new rom is loaded. If you record a video while the game is not displaying, it does show correctly in the recorded video. You can also see the rom selection screen and pick from the recent roms with no issues.

Build was done on Arch Linux

**Comment by SourMesen** [comment]
Thanks for the report.  What do you see exactly? Just a black screen?  Or do you see the game selection screen while the game's audio is playing?
If you see a black screen, are you able to display stuff like the FPS counters on it, or is it always fully black no matter what you do? (e.g pause, fps counters, change emulation speed, etc.)

**Comment by McKayJT** [comment]
It's a blank, black screen. 
Most of the time it's just completely blank, but sometimes you can see a faded portion of the rom loaded message (see attached image). Pausing does not change the display. Changing emulation speed doesn't affect anything. If I put the FPS counter on, most of the time it's the same blank, black display, but sometimes it will show up like in the attached screenshot. 
![2018-01-22-042655_838x1048_scrot](https://user-images.githubusercontent.com/34872500/35205433-0f23f184-ff2d-11e7-95e7-f189aecc8437.png)

**Comment by SourMesen** [comment]
I haven't actually tried this on Arch Linux, but I haven't been able to reproduce this on 2 different computers running Ubuntu (one in a VM, the other from a USB boot drive).  I've probably launched the Linux build a 100+ times in the last few weeks, and this problem did not occur a single time) I know there are other people using Arch Linux too, so I'm assuming it's not a distribution-specific issue.

Any chance you could try building from the latest commit & seeing if this is still occurring on your end?
Maybe it's related to the window manager you're using?

**Comment by frakswe** [comment]
yes black screen here with mint 18.3 aswell.i've compiled it myself with the same results.
the only thing that i can think of is that mint 18.3 uses glibc 2.23 while you recommend 2.24+, but i'd figure that would only apply to the binary you supply for us?

tried with mint 19 on another computer and it worked as it should.

**Comment by SourMesen** [comment]
I actually had a Linux Mint 18.3 VM already so I just gave it a try, but it seems to work as expected.
Are there any messages in Tools->Log Window that might look like an error?
Does power cycling fix the problem like in the original report, or do you only ever get a black screen?

glibc on my VM is 2.23 (and actually seems to work with the official builds, so I'm not sure why 2.24 is listed as the requirement), so this is probably unrelated.

Which window manager are you running?  I'm using Cinnamon on my end.

**Comment by frakswe** [comment]
errors i get in the log are "/dev/input" permissions.
i only get picture when i do "poweroff" in the menu, it then shows a snapshot picture of the game.

i'm using Cinnamon aswell.

**Comment by frakswe** [comment]
is there a way to get some realtime debug output from mono?

**Comment by SourMesen** [comment]
It sounds like this might be a different issue that the original one.
Do you get any sound at all while the game is running?

From what you've told me, it lookslike the SDL library is failing to properly initialize the video renderer, but it doesn't seem like it's returning any error, for some reason.

I think I've seen this before on a VM where hardware acceleration was disabled.
You can try confirming this on your end by changing this line: https://github.com/SourMesen/Mesen/blob/master/Linux/SdlRenderer.cpp#L58

Change it from:
```_sdlRenderer = SDL_CreateRenderer(_sdlWindow, -1, baseFlags | SDL_RENDERER_ACCELERATED);```
to
```_sdlRenderer = SDL_CreateRenderer(_sdlWindow, -1, baseFlags | SDL_RENDERER_SOFTWARE);```

If hardware acceleration is the problem, this should fix the video output.  The video output might be somewhat choppy with hardware acceleration disabled, though.  If changing this fixes your problem, you might want to check if your opengl drivers are working properly, or try updating them, etc.

**Comment by frakswe** [comment]
yes i get sound. i actually suspected the sdl renderer aswell but figured that i would notice other games and apps acting strange aswell. i'll get back to you with the result.

**Comment by frakswe** [comment]
sadly still black screen.

**Comment by frakswe** [comment]
i'll play around some with sdl renderer examples, maybe i'll figure out what's going on. thanks anyway.

**Comment by indexitis** [comment]
I experience this issue on Debian testing. I'm using the official release, didn't try building it myself yet. Just a power cycle won't fix it for me. I have to change the video size and then power cycle, then video works fine. By changing the video size I mean that the dimensions of the window must change somehow. Either by altering options > video size or simply dragging to expand the window.

So load game > change video size > power cycle. It can also work with change video size > load game > power cycle. This black screen also reappears whenever I change the video size during gameplay, and needs another power cycle to work again.

I only need to do this once every time the program is launched (unless the video size changes again). I can then switch games and it will work fine without needing any more messing around.

**Comment by SourMesen** [comment]
@indexitis  Thanks for the info - could you give me the version of Debian as well as the window manager you're using?  I'll try to build a matching Debian VM when I get a chance to see if I can reproduce it.

**Comment by indexitis** [comment]
@SourMesen It's a pretty recent install. I installed stable (Debian stretch) a few days ago via netinst iso and then upgraded to testing (buster), following the instructions from [this page](https://wiki.debian.org/DebianTesting). I am using KDE Plasma 5.13.2. Nvidia proprietary drivers as well if that helps.

cat /proc/version:
Linux version 4.17.0-1-amd64 (debian-kernel@lists.debian.org) (gcc version 7.3.0 (Debian 7.3.0-26)) #1 SMP Debian 4.17.8-1 (2018-07-20)

**Comment by indexitis** [comment]
Seem to have same problem, with same solution, running Mint 18.3 with Cinnamon on a separate machine. So don't think it's a KDE problem.

Like the other one, this machine also has Nvidia proprietary drivers. Don't know if that's relevant yet or just a coincidence. Messing around with VirtualBox now... Will get back with results in a while.

**Comment by indexitis** [comment]
Issue does not occur with Intel GPU and i915 driver. Therefore I now believe this problem (at least on my system) is related to both Linux and one of:

1. Nvidia proprietary drivers, or
2. Nvidia graphics cards in general

I have no way of testing AMD or other cards, but the issue does occur on two separate Nvidia cards (different models as well), both of which were using proprietary drivers. I will try to narrow it down further but this at least seems to explain why it couldn't be reproduced on VMs. Would be great if others experiencing this issue could mention their graphics hardware and drivers as well.

**Comment by indexitis** [comment]
Issue not present for me when using nouveau. The only unifying factors I've found are that this issue is present when the proprietary Nvidia drivers are used on Linux. May affect other cards as well. Don't think I can go any further with this on my own. I'm happy to help test anything though.

**Comment by frakswe** [comment]
I can confirm it works with amd and linux mint 19,for what it's worth.

**Comment by SourMesen** [comment]
@frakswe  I'm not overly familiar with signals in Linux, but from a bit of googling, it sounds like Mono uses SIGPWR internally for a number of things.  Looks like they recommend ignoring SIGPWR altogether when debugging Mono apps with gdb: https://www.mono-project.com/docs/debug+profile/debug/#debugging-with-gdb

**Comment by frakswe** [comment]
ok that certainly suppressed the hits,only get hits then entering the menu now:

Thread 1 "mono" received signal SIGSEGV, Segmentation fault.
0x0000000040cd98a5 in ?? ()
(gdb) backtrace
#0  0x0000000040cd98a5 in  ()
#1  0x00007f9e99943730 in  ()
#2  0x000000004120d6fb in  ()
#3  0x0000000041247c10 in  ()
#4  0x0000000000000000 in  ()
(gdb) c

fairly useless output...

**Comment by frakswe** [comment]
i'm ready to make a post at https://devtalk.nvidia.com/default/board/98/linux/ , but it's your program and i don't wanna be too rash about it.

**Comment by SourMesen** [comment]
Ok, so I've spent a few hours doing some testing and have been able to reproduce the issue on my end, but unfortunately haven't been able to find a reliable solution as of yet.

I tested off a "Try Ubuntu 16" usb stick, installed mono, SDL2, downloaded 0.9.6 and tried running it - and it worked just fine, no issues, like my VM.
Then I installed nvidia's drivers (instead of whatever default driver the usb key was using), restarted Xorg and ran 0.9.6 again, black screen.  Changing scale and power cycling (usually) fixes the problem.

I tried adding some extra debugging to the SDL code, and it doesn't look like any of the SDL calls fail.
I also managed to narrow down which part of the power cycle code causes it to start working again, but I have not been able to narrow it down as far as I would like, and I can't make any sense of it yet (the part of the power cycle code that fixes it should have no actual real impact on the SDL/rendering code..)

In my testing I've seen some pretty weird issues, at times, when I had a black screen, clicking outside the window (e.g to focus on another program) made the game's screen appear (and I haven't been able to understand how that could happen, either)

So, basically, it seems like Nvidia's Linux drivers are a factor in this case.  But it's still also possible that my code or SDL's code is somehow doing something wrong, which is causing the issue.  It might be worth testing with older nvidia drivers (or running older versions of Mesen) to see if it fixes the issue.

**Comment by indexitis** [comment]
I'll start doing some testing then. Hopefully can get back with results in a little while.

**Comment by indexitis** [comment]
Issue is present in Mesen 0.6.1, which is apparently the first release to support Linux. Tested a bunch of random versions in between with the same result. Seems like I don't need to resize the window anymore though - not even on the version that I previously had to. Weird.

**Comment by indexitis** [comment]
Nvidia drivers 384.130, 390.77, and 396.54 all have the issue. It's kind of a pain for me to test earlier than 384 so I'm gonna hold off that for now. My hardware also doesn't support 340 and earlier. Maybe @frakswe it's time to make that Nvidia forums post.

**Comment by andre-paris** [comment]
As a workaround you could add a command line option to force an extra Power Cycle after the rom is loaded.
This would help significantly people like me who use Mesen to develop nes games on Linux.

**Comment by scottlinux** [comment]
I use Mesen on Fedora 28, AMD graphics (RX 480), with the open source amdgpu driver. All works great for me and no issues with a black screen. 

Mesen version: 0.9.6

Cheers,

**Comment by Silverfoenix** [comment]
I had to sign in to give some sort of **hint**
I'm using the latest development build .97
I'm a total noob regarding these matters.. however I noticed No Video on my laptop (i5 - Ati mobilty) as well, Only audio is running!
When pausing I can see Crosshair of the mouse (Playing Duck hunt) but resuming ingame nothing..
I tried **applying a filter and it worked!** Only the **NTSC Bisquit** ones so far provides Video!!!
Also the Crosshair disapears when resuming.

**Comment by spiiin** [comment]
I got a similar error while building on windows 10. I tried different versions of the windows sdks and update drivers to latest, but nothing helps - I've got only black screen and audio. Power/reset emulator doesn't help. NTSC Bisquit filter provides video for me too.

My configuration - nvidia geforce gtx 960m / intel hd graphics 5600 (I tried to run on both cards - both not worked), diriver - 419.35-notebook-win10-64bit-international-whql drivers (latest for this card)

**Comment by SourMesen** [comment]
If Bisqwit's filter "solves" the issue, it's most likely only because it ends up changing the rendering resolution.
If you're getting a black screen with the builds you compile on Windows, but not with the appveyor builds, then the problem is most likely caused by the build itself.  (I'm assuming you're building the DirectX version and not trying to run the SDL code under Windows, either)
If the intel graphics cause a black screen, this is most likely a different issue than the Linux nvidia problem.

**Comment by SourMesen** [comment]
@spiiin The issue you were getting on Windows should be fixed as of the latest commit

**Comment by ThothXV** [comment]
>(the part of the power cycle code that fixes it should have no actual real impact on the SDL/rendering code..)

What part of the power cycle code was it? @SourMesen

**Comment by SourMesen** [comment]
@ThothXV  I can't say I remember, unfortunately, been a long time since I last looked into this.  Essentially it looks like the SDL video backend somehow doesn't initialize itself properly on the first try (when using the nvidia drivers).  Whether or not this is due to my code, SDL, Mono, or nvidia I still do not have an answer to (but it only happens when using nvidia's drivers).

Power cycling in general shouldn't have an impact on the video backend at all, so that was what my comment referred to (but it's possible some event occurs during power cycling which causes the SDL backend to reinit itself).  Essentially all the solutions (change scale, change video filter, etc.) work because they cause the video backend to be reinitialized from scratch, which apparently works properly the 2nd time.

**Comment by jackwilsdon** [comment]
It appears that restarting the video decoder thread fixes the problem, although I have absolutely no idea why. This is the reason that power cycling the console works — it also restarts the video decoder thread.

I have confirmed it is the video decoder thread by calling `videoDecoder->StopThread()` followed by `videoDecoder->StartThread()` using `gdb`, and this immediately brings the picture back.

I hope this is a good starting place for someone to work out how restarting the thread is fixing the issue (and to find the root cause).

I have already tried setting all of the variables that are set within `StopThread` and `StartThread` line by line using `gdb`'s `call`, however this also didn't fix the problem.

**Comment by Sembiance** [comment]
So this also happens to me. I'm running `Gentoo Linux` with `nvidia-drivers 430.09` and `mono 5.20.1.19` and the latest Mesen Github master compiled manually from source. I tried compiling with both g++ and clang, both produce the same results, the video is black but the audio works fine. Switching in the code to `SDL_RENDERER_SOFTWARE` did not fix it.  Like other reports above, sometimes I can get the video to come back when playing around with filters and such.
As a test I opened nvidia-settings and in OpenGL Settings I tried disabling `Sync to VBlank` `Allow Flipping` and `Allow G-SYNC` and none of them helped. Neither did toggling V-Sync in Messen.

**Comment by SeanRamey** [comment]
I am experiencing this issue as well. Arch Linux with proprietary Nvidia drivers. Also, the filters do not seem to work for me. They just give a black screen and power cycling doesn't help. As soon as I select no filter and power cycle I get video again.

**Comment by lamskoy** [comment]
Similar issue on ArchLinux:
Proprietary drivers: nvidia-dkms 430.40
Kernel: 5.2.5-arch1-1-ARCH #1 SMP PREEMPT 
I set screen size to 1x, disable all filters and do power cycle. 
Then I get video

**Comment by severinkaderli** [comment]
The same issue happens for on ArchLinux.

Drivers: nvidia 430.40  
Kernel: 5.2.6-arch1-1-ARCH #1 SMP PREEMPT

**Comment by sircodalotkob** [comment]
I wonder if it could have something to do with the OpenGL context ownership?

The Mono user interface is driven by the application's main thread whereas the OpenGL rendering is done by another thread, spawned for that job. There may be ownership/synchronizations issues.

I came across the idea because I also get a crash when resizing the main window with a game running in it:

=================================================================
        Native Crash Reporting
=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================
/proc/self/maps:
40195000-40244000 rw-p 00000000 00:00 0 
4032a000-4033a000 rwxp 00000000 00:00 0 
40eef000-40ef1000 r-xs 00000000 00:2d 48                                 /tmp/.glhUEaIE (deleted)
4156f000-418ab000 rwxp 00000000 00:00 0 
418ab000-4195a000 rw-p 00000000 00:00 0 
41e14000-41ec3000 rw-p 00000000 00:00 0 
5654abc06000-5654abc32000 r--p 00000000 08:12 1329068                    /usr/bin/mono-sgen
5654abc32000-5654abf58000 r-xp 0002c000 08:12 1329068                    /usr/bin/mono-sgen
5654abf58000-5654ac09a000 r--p 00352000 08:12 1329068                    /usr/bin/mono-sgen
5654ac09b000-5654ac0a2000 r--p 00494000 08:12 1329068                    /usr/bin/mono-sgen
5654ac0a2000-5654ac0a7000 rw-p 0049b000 08:12 1329068                    /usr/bin/mono-sgen
5654ac0a7000-5654ac0be000 rw-p 00000000 00:00 0 
5654ad7c2000-5654af4c4000 rw-p 00000000 00:00 0                          [heap]
7f2a58000000-7f2a595d1000 rw-p 00000000 00:00 0 
7f2a595d1000-7f2a5c000000 ---p 00000000 00:00 0 
7f2a5c000000-7f2a5c509000 rw-p 00000000 00:00 0 
7f2a5c509000-7f2a60000000 ---p 00000000 00:00 0 
7f2a63000000-7f2a63100000 rw-p 00000000 00:00 0 
7f2a6317c000-7f2a631fc000 rw-p 00000000 00:00 0 
7f2a63200000-7f2a63400000 rw-s 00000000 00:06 497                        /dev/nvidiactl
7f2a63400000-7f2a63800000 rw-s 00000000 00:06 497                        /dev/nvidiactl
7f2a63800000-7f2a63c00000 rw-s 00000000 00:06 497                        /dev/nvidiactl
7f2a63c00000-7f2a63e00000 rw-s 00000000 00:06 497                        /dev/nvidiactl
7f2a63e00000-7f2a63f00000 rw-p 00000000 00:00 0 
7f2a63f00000-7f2a64000000 rw-s 00000000 00:06 497                        /dev/nvidiactl

=================================================================The Mono user interface is driven by the application's main thread whereas the OpenGL rendering is 
        Basic Fault Adddress Reporting
=================================================================
Memory around native instruction pointer (0x7f2ae41499f7):0x7f2ae41499e7  9f f0 09 00 00 48 8b 93 18 01 00 00 48 8d 42 50  .....H......H.BP
0x7f2ae41499f7  48 8b 52 30 44 8b b8 ec 0e 00 00 44 8b b0 e8 0e  H.R0D......D....
0x7f2ae4149a07  00 00 48 89 14 24 4c 8b 6f 68 48 8b 7b 40 83 bf  ..H..$L.ohH.{@..
0x7f2ae4149a17  00 01 00 00 13 0f 86 a3 00 00 00 4c 8b 25 8f da  ...........L.%..

=================================================================
        Native stacktrace:
=================================================================
        0x5654abd3300a - mono : (null)
        0x5654abcc117e - mono : (null)
        0x5654abc33828 - mono : (null)
        0x7f2af9879930 - /usr/lib/libpthread.so.0 : (null)
        0x7f2ae41499f7 - /usr/lib/libGLX_nvidia.so.0 : (null)

=================================================================
        Telemetry Dumper:
=================================================================
Thread 0x7f2ae52ff700 may have been prematurely finalized* Assertion at mono-threads.c:634, condition `info' not met
malloc(): unsorted double linked list corrupted

**Comment by SourMesen** [comment]
So it nearly took 2 whole years, but I finally had the time to sit down, setup a real Linux environment and debug/fix the problem.
As of ba79209bd5d7d8f8a7d810f4cd94cf5be8e00d91, this should be fixed.  If you still have issues with it when using the latest appveyor builds, let me know.
